### DISCUSSION TITLE
Populating Nested Classes with Mapper

### DISCUSSION THREAD

**Discussion by toddbanister** [sqlalchemy/sqlalchemy#5958] [Usage Questions]
I am just beginning to learn sqlalchemy so I apologize for such a basic question. I am trying to understand how to map table columns into my domain model classes that have nested classes within them.

For example, I have a Weapon class that is defined with a nested Effects class like:
```
        class Effects():
            def __init__(self, unboxed: int = 0, boxed: int = 0, unboxed_red: bool = False):
                self.boxed: int = boxed
                self.unboxed: int = unboxed
                self.unboxed_red: bool = unboxed_red

        class Weapon():
            def __init__(self, name: str = ""):
                self.effect: Effects = Effects()
                self.name: str = name
```

In my sqlite db, all of the data is stored within a single table. Thus, I created my sqlalchemy table definition as follows:

```
        t_Cards_Weapon = Table(
            'Cards_Weapon', metadata,
            Column('Name', Numeric(50), primary_key=True, nullable=False),
            Column('Boxed', Integer, server_default=text("0")),
            Column('Unboxed', Integer),
            Column('Unboxed_Red', Boolean, server_default=text("FALSE"))
        )
```

But I am not sure how I can map the table columns to the nested Effects class using the sqlalchemy mapper:

```
            mapper(Weapon, t_Cards_Weapon, properties={
                'name': t_Cards_Weapon.c.Name,
                'boxed': t_Cards_Weapon.c.Boxed,
                'unboxed': t_Cards_Weapon.c.Unboxed,
                'unboxed_red': t_Cards_Weapon.c.Unboxed_Red
            })
```

Obviously, the mapper as currently designed only produces a "flat" Weapon's class and not the desired populated nested Effects class.

Can someone give me some direction on how to get started here?

Thanks for your time,
TB

**Comment by zzzeek** [comment]
it seems like you are looking at [single table inheritance](https://docs.sqlalchemy.org/en/14/orm/inheritance.html#single-table-inheritance) which you would map as follows (note I am using SQLAlchemy 1.4.0b3's notion of classical mapping, overall I would recommend using 1.4 for new projects as it will be released within a few weeks.)


```python

from sqlalchemy import Boolean
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import Numeric
from sqlalchemy import select
from sqlalchemy import Table
from sqlalchemy import text
from sqlalchemy.orm import registry


reg = registry()


t_Cards_Weapon = Table(
    "Cards_Weapon",
    reg.metadata,
    Column("Name", Numeric(50), primary_key=True, nullable=False),
    Column("Boxed", Integer, server_default=text("0")),
    Column("Unboxed", Integer),
    Column("Unboxed_Red", Boolean, server_default=text("FALSE")),
)


class Effects:
    def __init__(
        self, unboxed: int = 0, boxed: int = 0, unboxed_red: bool = False
    ):
        self.boxed: int = boxed
        self.unboxed: int = unboxed
        self.unboxed_red: bool = unboxed_red


class Weapon(Effects):
    def __init__(self, name: str = ""):
        self.effect: Effects = Effects()
        self.name: str = name


reg.map_imperatively(Effects, t_Cards_Weapon)
reg.map_imperatively(Weapon, inherits=Effects)

print(select(Weapon))
```


what you may want to consider is some means of disambiguating rows in your table that indicate an Effect or a Weapon object.   It's likely possible to use the presence of "name" as this indicator but SQLAlchemy prefers you establish a distinct [discriminator column](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=polymorphic_on#sqlalchemy.orm.mapper.params.polymorphic_on).
